### ISSUE TITLE
Replace List<T> with History<T>

### ISSUE THREAD

**Issue by Zhuinden** [Zhuinden/simple-stack#82]
Just like square/flow, it could make sense to provide a custom ~~Iterable~~ `List<T>` class.

that way, HistoryBuilder could move to `History.Builder`.

What's the benefit? Instead of having to remember things like:

    stateChange.getNewState().get(size-1)`

I could also add `fromTop(0)` and `top()` like I did to Backstack.

That way, `backstack.top()` and `fromTop()` could actually come from `History` interface! Is that a hack at that point? Not sure. `backstack.getHistory()` could just return a `History` instead of the `List<T>` it does now.

Or I could remove `backstack.top()` and `backstack.fromTop()`, move it to `History`, and then one can do `backstack.getHistory()` which would have these methods.

Or just keep both.

----

But is this too intrusive? Is it worth it? Kotlin can already access `last()` and stuff from a `List`, so this is more-so something the Java world would benefit from.

**Comment by Zhuinden** [comment]
Actually, `spliterator` is also in `Iterable<T>`.

How the heck does `square/flow` circumvent this? :smile:

**Comment by Zhuinden** [comment]
Answer:

```

**Comment by Zhuinden** [comment]
Actually...

`public class History<T> extends AbstractList<T> {`

This might be the solution.

**Comment by Zhuinden** [comment]
First draft is failure due to too many generics, this is a save-out before I revert again.

``` java
package com.zhuinden.simplestack;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.AbstractList;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;

/**
 * An immutable wrapper over backstack history with some additional helper methods.
 */
public class History<T> extends AbstractList<T> implements List<T> {
    private final List<T> elements;
    
    History() {
        this(Collections.<T>emptyList());
    }
    
    History(List<T> elements) {
        this.elements = Collections.unmodifiableList(elements);
    }


    /**
     * Creates a new history from the provided keys.
     *
     * @param keys the provided keys
     * @param <T> possible base type of the keys
     * @return the history
     */
    @SuppressWarnings("unchecked") // @SafeVarargs is API 19+
    @NonNull
    public static <T> History<? super T> from(T... keys) {
        return builderFrom(Arrays.asList(keys)).build();
    }

    /**
     * Creates a new history from the provided keys.
     *
     * @param keys the provided keys
     * @param <T> possible base type of the keys
     * @return the history
     */
    @NonNull
    public static <T> History<T> from(@NonNull List<? extends T> keys) {
        return builderFrom(keys).build();
    }

    /**
     * Creates a new history builder based on the {@link Backstack}'s history.
     *
     * @param backstack the {@link Backstack}.
     * @return the newly created {@link Builder}.
     */
    @NonNull
    public static <T> Builder<? super T> builderFrom(@NonNull Backstack backstack) {
        if(backstack == null) {
            throw new IllegalArgumentException("Backstack cannot be null!");
        }
        History<? super T> history = backstack.getHistory();
        return newBuilder().addAll(history);
    }

    /**
     * Creates a new history builder based on the {@link BackstackDelegate}'s managed backstack history.
     *
     * @param backstackDelegate the {@link BackstackDelegate}.
     * @return the newly created {@link Builder}.
     */
    @NonNull
    public static <T> Builder<? super T> builderFrom(@NonNull BackstackDelegate backstackDelegate) {
        if(backstackDelegate == null) {
            throw new IllegalArgumentException("BackstackDelegate cannot be null!");
        }
        return builderFrom(backstackDelegate.getBackstack());
    }

    /**
     * Creates a new history builder builderFrom the provided ordered elements.
     *
     * @param keys
     * @return the newly created {@link Builder}.
     */
    @SuppressWarnings("unchecked") // @SafeVarargs is API 19+
    @NonNull
    public static <T> Builder<? super T> builderFrom(T... keys) {
        return builderFrom(Arrays.asList(keys));
    }

    /**
     * Creates a new history builder builderFrom the provided ordered collection.
     *
     * @param keys
     * @return the newly created {@link Builder}.
     */
    @NonNull
    public static <T> Builder<T> builderFrom(@NonNull List<? extends T> keys) {
        for(T key : keys) {
            if(key == null) {
                throw new IllegalArgumentException("Cannot provide `null` as a key!");
            }
        }
        return History.<T>newBuilder().addAll(keys);
    }

    /**
     * Creates a new empty history builder.
     *
     * @return the newly created {@link Builder}.
     */
    @NonNull
    public static <T> Builder<T> newBuilder() {
        return new Builder<>();
    }

    /**
     * Creates a new array list of object that contains only the provided key.
     *
     * @param key
     * @return an array list of object that contains the key.
     */
    @NonNull
    public static <T> History<? super T> single(@NonNull T key) {
        return History.<T>newBuilder()
                .add(key)
                .build();
    }

    @Override
    public boolean add(T t) {
        return elements.add(t);
    }

    @Override
    public T set(int index, T element) {
        return elements.set(index, element);
    }

    @Override
    public void add(int index, T element) {
        elements.add(index, element);
    }

    @Override
    public T remove(int index) {
        return elements.remove(index);
    }

    @Override
    public int indexOf(Object o) {
        return elements.indexOf(o);
    }

    @Override
    public int lastIndexOf(Object o) {
        return elements.lastIndexOf(o);
    }

    @Override
    public void clear() {
        elements.clear();
    }

    @Override
    public boolean addAll(int index, Collection<? extends T> c) {
        return elements.addAll(index, c);
    }

    @NonNull
    @Override
    public Iterator<T> iterator() {
        return elements.iterator();
    }

    @NonNull
    @Override
    public ListIterator<T> listIterator() {
        return elements.listIterator();
    }

    @NonNull
    @Override
    public ListIterator<T> listIterator(int index) {
        return elements.listIterator(index);
    }

    @NonNull
    @Override
    public List<T> subList(int fromIndex, int toIndex) {
        return elements.subList(fromIndex, toIndex);
    }

    @Override
    public boolean equals(Object o) {
        if(o == null) {
            return false;
        }
        if(!(o instanceof History)) {
            return false;
        }
        return elements.equals(((History)o).elements);
    }

    @Override
    public int hashCode() {
        int result = super.hashCode();
        result = 31 * result + elements.hashCode();
        return result;
    }

    @Override
    public boolean isEmpty() {
        return elements.isEmpty();
    }

    @Override
    public boolean contains(Object o) {
        return elements.contains(o);
    }

    @NonNull
    @Override
    public Object[] toArray() {
        return elements.toArray();
    }

    @NonNull
    @Override
    public <T1> T1[] toArray(@NonNull T1[] a) {
        return elements.toArray(a);
    }

    @Override
    public boolean remove(Object o) {
        return elements.remove(o);
    }

    @Override
    public boolean containsAll(@NonNull Collection<?> c) {
        return elements.containsAll(c);
    }

    @Override
    public boolean addAll(@NonNull Collection<? extends T> c) {
        return elements.addAll(c);
    }

    @Override
    public boolean removeAll(@NonNull Collection<?> c) {
        return elements.removeAll(c);
    }

    @Override
    public boolean retainAll(@NonNull Collection<?> c) {
        return elements.retainAll(c);
    }

    @Override
    public String toString() {
        return Arrays.toString(elements.toArray());
    }

    @Override
    public T get(int index) {
        return elements.get(index);
    }

    @Override
    public int size() {
        return elements.size();
    }

    /**
     * Class for creating a history of items, for backstack history.
     */
    public static class Builder<T>
            implements Iterable<T> {
        private ArrayList<T> list = new ArrayList<>();

        private Builder() { // use newBuilder()
        }

        /**
         * Adds the keys to the builder.
         *
         * @param keys
         * @return the current builder.
         */
        @NonNull
        public Builder<T> addAll(@NonNull List<? extends T> keys) {
            if(keys == null) {
                throw new IllegalArgumentException("Provided collection cannot be null");
            }
            this.list.addAll(keys);
            return this;
        }

        /**
         * Adds the keys to the builder at a given index.
         *
         * @param keys
         * @param index
         * @return the current builder.
         */
        @NonNull
        public Builder<T> addAllAt(@NonNull List<? extends T> keys, int index) {
            if(keys == null) {
                throw new IllegalArgumentException("Provided collection cannot be null");
            }
            this.list.addAll(index, keys);
            return this;
        }

        /**
         * Clears the history builder.
         *
         * @return the current builder.
         */
        @NonNull
        public Builder<T> clear() {
            list.clear();
            return this;
        }

        /**
         * Returns if the given key is contained within the builder.
         *
         * @param key
         * @return true if the builder contains the given key.
         */
        public boolean contains(@NonNull T key) {
            checkKey(key);
            return list.contains(key);
        }

        /**
         * Returns if the builder contains all provided keys.
         *
         * @param keys
         * @return true if the builder contains all keys.
         */
        public boolean containsAll(@NonNull Collection<? extends T> keys) {
            if(keys == null) {
                throw new IllegalArgumentException("Keys cannot be null!");
            }
            return list.containsAll(keys);
        }

        /**
         * Returns the size of the builder.
         *
         * @return the number of keys in the builder.
         */
        public int size() {
            return list.size();
        }

        /**
         * Removes the given key builderFrom the builder.
         *
         * @param key
         * @return the current builder.
         */
        @NonNull
        public Builder<T> remove(@NonNull T key) {
            checkKey(key);
            list.remove(key);
            return this;
        }

        /**
         * Remove the key at the given index.
         *
         * @param index
         * @return the current builder.
         */
        @NonNull
        public Builder<T> removeAt(int index) {
            list.remove(index);
            return this;
        }

        /**
         * Removes all keys builderFrom the builder not contained inside the provided keys.
         *
         * @param keys
         * @return the current builder.
         */
        @NonNull
        public Builder<T> retainAll(@NonNull Collection<? extends T> keys) {
            checkKeys(keys);
            list.retainAll(keys);
            return this;
        }

        /**
         * Returns if the builder is empty.
         *
         * @return true if the builder does not contain any keys
         */
        public boolean isEmpty() {
            return list.isEmpty();
        }

        /**
         * Removes the last entry in the builder.
         * If the builder is empty, an exception is thrown.
         *
         * @return the current builder.
         */
        @NonNull
        public Builder<T> removeLast() {
            if(list.isEmpty()) {
                throw new IllegalStateException("Cannot remove element builderFrom empty builder");
            }
            list.remove(list.size() - 1);
            return this;
        }

        /**
         * Removes all keys until the provided key is found.
         * If the key is not found, an exception is thrown.
         *
         * @param key
         * @return the current builder.
         */
        @NonNull
        public Builder<T> removeUntil(@NonNull T key) {
            checkKey(key);
            while(!list.isEmpty() && !key.equals(getLast())) {
                removeLast();
            }
            if(list.isEmpty()) {
                throw new IllegalArgumentException("[" + key + "] was not found in history!");
            }
            return this;
        }

        /**
         * Returns the index of the provided key.
         *
         * @param key
         * @return the index, -1 if not found.
         */
        public int indexOf(@NonNull T key) {
            checkKey(key);
            return list.indexOf(key);
        }

        /**
         * Returns the key at the given index.
         *
         * @param index
         * @return the key at the given index
         */
        @NonNull
        public <K> K get(int index) {
            // noinspection unchecked
            return (K) list.get(index);
        }

        /**
         * Returns the last element of the builder.
         * If the builder is empty, null is returned.
         *
         * @return the key at the last index
         */
        @Nullable
        public <K> K getLast() {
            // noinspection unchecked
            return (K)(list.isEmpty() ? null : list.get(list.size() - 1));
        }

        /**
         * Adds the provided key as the last element of the builder.
         *
         * @param key
         * @return the current builder.
         */
        @NonNull
        public Builder<T> add(@NonNull T key) {
            checkKey(key);
            list.add(key);
            return this;
        }

        /**
         * Adds the provided key at the provided index.
         *
         * @param key
         * @param index
         * @return the current builder.
         */
        @NonNull
        public Builder<T> add(@NonNull T key, int index) {
            checkKey(key);
            list.add(index, key);
            return this;
        }

        /**
         * Provides an iterator for the builder.
         *
         * @return the iterator
         */
        @NonNull
        @Override
        public Iterator<T> iterator() {
            return list.iterator();
        }

        /**
         * Creates a copied version of the builder as an ArrayList.
         *
         * @return the built history.
         */
        @NonNull
        public History<T> build() {
            return new History<>(this.list);
        }

        // validations
        private void checkKey(Object key) {
            if(key == null) {
                throw new IllegalArgumentException("History key cannot be null!");
            }
        }

        private void checkKeys(Collection<? extends T> keys) {
            if(keys == null) {
                throw new IllegalArgumentException("Keys cannot be null!");
            }
        }
    }
}
```

**Comment by Zhuinden** [comment]
I can actually add this in `1.9` as long as the `History extends ArrayList<T>` for now, instead of `AbstractList<T>`.

Then again, the change between 1.0.0 and 1.1.0 was much more intrusive :D

**Comment by Zhuinden** [comment]
So yeah, with that in mind, `History` is properly `AbstractList`.

If HistoryBuilder is kept where it is, then most thing is just deprecation, it actually works quite nicely.

Additional effect is the resolution of some generic iffiness which forced the usage of HistoryBuilder.whatever(), but wasn't really necessary.

**Comment by Zhuinden** [comment]
Released in 1.9.0
